libobs/audio-monitoring/nix/pulseaudio-output.c

```diff
@@ -0,0 +1,487 @@
+	case 1:
+		process_byte(resample_data[0], resample_frames,
+		             monitor->channels, vol);
+		break;
+	case 2:
+		process_short(resample_data[0], resample_frames,
+		              monitor->channels, vol);
+		break;
+	default:
+		process_float(resample_data[0], resample_frames,
+		              monitor->channels, vol);
+		break;
+	}
+}
+
+static void do_stream_write(void *param)
+{
+	PULSE_DATA(param);
+	uint8_t *buffer = NULL;
+
+	while (data->new_data.size >= data->buffer_size &&
+	       data->bytesRemaining > 0) {
+		size_t bytesToFill = data->buffer_size;
+
+		if (bytesToFill > data->bytesRemaining)
+			bytesToFill = data->bytesRemaining;
+
+		pa_stream_begin_write(data->stream, (void **) &buffer,
+		                                &bytesToFill);
+
+		circlebuf_pop_front(&data->new_data, buffer, bytesToFill);
+
+		pulseaudio_lock();
+		pa_stream_write(data->stream, buffer, bytesToFill, NULL,
+		                      0LL,
+		                      PA_SEEK_RELATIVE);
+		pulseaudio_unlock();
+
+		data->bytesRemaining -= bytesToFill;
+	}
+}
+
+static void on_audio_playback(void *param, obs_source_t *source,
+                              const struct audio_data *audio_data, bool muted)
+{
+	struct audio_monitor *monitor = param;
+	float vol = source->user_volume;
+	size_t bytes;
+
+	uint8_t *resample_data[MAX_AV_PLANES];
+	uint32_t resample_frames;
+	uint64_t ts_offset;
+	bool success;
+
+	if (pthread_mutex_trylock(&monitor->playback_mutex) != 0) {
+		return;
+	}
+	if (os_atomic_load_long(&source->activate_refs) == 0) {
+		goto unlock;
+	}
+
+	success = audio_resampler_resample(monitor->resampler, resample_data,
+	                                   &resample_frames, &ts_offset,
+	                                   (const uint8_t *const *) audio_data->data,
+	                                   (uint32_t) audio_data->frames);
+	if (!success) {
+		goto unlock;
+	}
+
+	bytes = monitor->bytes_per_frame * resample_frames;
+
+	if (muted) {
+		memset(resample_data[0], 0, bytes);
+	} else {
+		if (!close_float(vol, 1.0f, EPSILON)) {
+			process_volume(monitor, vol, resample_data,
+			               resample_frames);
+		}
+	}
+
+	circlebuf_push_back(&monitor->new_data, resample_data[0], bytes);
+	monitor->packets++;
+	monitor->frames += resample_frames;
+
+unlock:
+	pthread_mutex_unlock(&monitor->playback_mutex);
+	do_stream_write(param);
+}
+
+static void
+pulseaudio_stream_write(pa_stream *p, size_t nbytes, void *userdata)
+{
+	UNUSED_PARAMETER(p);
+	PULSE_DATA(userdata);
+
+	pthread_mutex_lock(&data->playback_mutex);
+	data->bytesRemaining += nbytes;
+	pthread_mutex_unlock(&data->playback_mutex);
+
+	pulseaudio_signal(0);
+}
+
+static void pulseaudio_server_info(pa_context *c, const pa_server_info *i,
+                                   void *userdata)
+{
+	UNUSED_PARAMETER(c);
+	UNUSED_PARAMETER(userdata);
+
+	blog(LOG_INFO, "Server name: '%s %s'",
+	     i->server_name, i->server_version);
+
+	pulseaudio_signal(0);
+}
+
+static void
```

Comment:
Style:  Don't line break between the function return type and the function name in a function declaration.